docs/developers_guide/FlinkComponents.md

```diff
@@ -1,22 +1,72 @@
 # Flink components
- 
-Sources, sinks and custom transformations are based on 
-Flink API. Access to various helpers is provided by 
-`FlinkCustomNodeContext`. Special care should be taken to handle
-- lifecycle
-- exception handling properly
+
+Sources, sinks and custom transformations are based on
+Flink API. Access to various helpers is provided by
+`FlinkCustomNodeContext`. Special care should be taken to handle:
+- lifecycle - preparing the operator (source, sink or functions registered by custom transformers), closing resources, handling failures and restoring the state.
+  See Flink's operators lifecycle for more details [here](https://ci.apache.org/projects/flink/flink-docs-master/docs/internals/task_lifecycle/)
+- exceptions, e.g. during deserialization, since any thrown and unhandled exception by the source, causes the Flink job to restart.
 
 :warning: **Flink components should not extend Lifecycle** - it won't be handled properly
 
 ## Sources
```

Comment:
What about checkpointing? Do we need to write something about it? 

In case of Kafka source I think Nu's Kafka source records Kafka offsets during the checkpoint and restarts from this offset.